### PR TITLE
[DOCS] Update subcurrency example in introductory section

### PR DESCRIPTION
**Pull request by ChrisChinchilla**
### Description

Following https://github.com/ethereum/solidity/pull/6566 this PR polishes the next section of the "introduction to smart contracts" page to make it flow better for people following the tutorial. As part of this I restructured the tutorial a fair bit.

### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [x] New tests have been created which fail without the change (if possible)
- [x] README / documentation was extended, if necessary
- [x] Changelog entry (if change is visible to the user)
- [x] Used meaningful commit messages

### CODE REVIEW

File: docs/introduction-to-smart-contracts.rst [hunk 1]
```diff
@@ -115,58 +108,56 @@ registering with username and password — all you need is an Ethereum keypair.
 
 This contract introduces some new concepts, let us go through them one by one.
 
-The line ``address public minter;`` declares a state variable of type address
```
Comment (chriseth) [inline docs/introduction-to-smart-contracts.rst:118]: Please do not change the line breaks here.

File: docs/introduction-to-smart-contracts.rst [hunk 2]
```diff
@@ -71,30 +71,23 @@ so that only you can alter the number.
...
-registering with username and password — all you need is an Ethereum keypair.
-
+registering with username and password, all you need is an Ethereum keypair.
 
 ::
 
     pragma solidity >=0.5.0 <0.7.0;
 
     contract Coin {
-        // The keyword "public" makes those variables
```
Comment (chriseth) [inline docs/introduction-to-smart-contracts.rst:87]: Don't you think these inline comments make it a bit more readable? With inline comments, you do not always have to flick back and forth and it is always clear (at least almost) what the text refers to.

File: docs/introduction-to-smart-contracts.rst (hunk 1 above)
Comment (ChrisChinchilla) [reply docs/introduction-to-smart-contracts.rst:118]: This was somewhat intentional, to break up the spacing and thus flow of a lot of explanation for someone new to Solidity, breaking each concept into a new paragraph, making it easier to scan through them one by one. White space helps with easing this flow. Unless I am misunderstanding what you mean?

File: docs/introduction-to-smart-contracts.rst (hunk 2 above)
Comment (ChrisChinchilla) [reply docs/introduction-to-smart-contracts.rst:87]: Personally I found it confusing that you see everything (without knowing what any of it is) and then you get the explanation, which for a beginner is a little confusing. Personally I had to go back and forth a lot with the current version to understand what lines meant, and re-read several times.

I have two ideas.

1. Add back the comments, but keep some of the further explanations after the code, offering people two ways to understand what the code does.
2. Change the order. Start with a blank contract, and add and explain each section piece by piece and then present the final code at the end. This is a more traditional way of structuring a getting started guide.

File: docs/introduction-to-smart-contracts.rst (hunk 1 above)
Comment (chriseth) [reply docs/introduction-to-smart-contracts.rst:118]: I mean the line breaks in the source code, not in the rendered output - it is hard to review this way.

File: docs/introduction-to-smart-contracts.rst (hunk 2 above)
Comment (chriseth) [reply docs/introduction-to-smart-contracts.rst:87]: I agree with you, but I think removing the comments exactly causes the problem that you see something without explanation ;)

Having said that, the exact comments might perhaps not be too helpful. Reading about "light clients" or things being "easily accessible from outside" does not really help people understand what this is about. We already know what state variables are, so this is good, but perhaps at least the comment on the constructor should stay? Perhaps you could also find better comments that explain in a single sentence what we are seeing here?

**Comment by ChrisChinchilla** [comment]
Try these recent updates @chriseth , added back comments (improved hopefully), plus some further explanation.

**Review by leonardoalt — DISMISSED** [review]
Lgtm

**Comment by ChrisChinchilla** [comment]
Thanks @leonardoalt ! Might just wait for an extra check from @chriseth as in my mind this intro content is very important.

File: docs/introduction-to-smart-contracts.rst [hunk 3]
```diff
@@ -71,40 +71,41 @@ so that only you can alter the number.
...
-        // This is the constructor whose code is
-        // run only when the contract is created.
+        // Constructor code is only run when the contract
+        // is created
         constructor() public {
             minter = msg.sender;
         }
 
+        // Sends an amount of newly created coins
+        // from the contract creator to an address
```
Comment (chriseth) [inline docs/introduction-to-smart-contracts.rst]: The contract creator does not really own them, so they are not sent from the contract creator.

Comment (chriseth) [reply docs/introduction-to-smart-contracts.rst]: Creates an amount of coins and assigns them to a specified address?

Comment (chriseth) [reply docs/introduction-to-smart-contracts.rst]: (can only be called by the contract creator)

File: docs/introduction-to-smart-contracts.rst [hunk 4]
```diff
@@ -115,58 +116,55 @@ registering with username and password — all you need is an Ethereum keypair.
...
-that does not allow any arithmetic operations. It is suitable for
-storing addresses of contracts or of keypairs belonging to external
-persons. The keyword ``public`` automatically generates a function that
-allows you to access the current value of the state variable
-from outside of the contract.
-Without this keyword, other contracts have no way to access the variable.
-The code of the function generated by the compiler is roughly equivalent
+The line ``address public minter;`` declares a state variable of type :ref:`address<address>`.
+The ``address`` type is a 160-bit value that does not allow any arithmetic operations.
+It is suitable for storing addresses of contracts, or of keypairs belonging to external persons.
```
Comment (chriseth) [inline docs/introduction-to-smart-contracts.rst]: keypairs are not stored, it is only a hash of the public part of a keypair that is stored.

File: docs/introduction-to-smart-contracts.rst [hunk 5]
```diff
@@ -115,58 +116,55 @@ registering with username and password — all you need is an Ethereum keypair.
...
-listen for those events being emitted on the blockchain without much
-cost. As soon as it is emitted, the listener will also receive the
-arguments ``from``, ``to`` and ``amount``, which makes it easy to track
-transactions. In order to listen for this event, you would use the following
-JavaScript code (which assumes that ``Coin`` is a contract object created via
-web3.js or a similar module)::
+an :ref:`"event" <events>`, which is emitted in the last line of the function
+``send``. Light Ethereum clients such as web applications can
+listen for these events emitted on the blockchain without much
+gas cost. As soon as it is emitted, the listener receives the
```
Comment (chriseth) [inline docs/introduction-to-smart-contracts.rst]:
```suggestion
cost. As soon as it is emitted, the listener receives the
```

File: docs/introduction-to-smart-contracts.rst [hunk 6]
```diff
@@ -179,36 +177,36 @@ web3.js or a similar module)::
         }
     })
 
-Note how the automatically generated function ``balances`` is called from
-the user interface.
+.. note::
```
Comment (chriseth) [inline docs/introduction-to-smart-contracts.rst]: I don't think this should be made into a note-block.

File: docs/introduction-to-smart-contracts.rst [hunk 7]
```diff
@@ -179,36 +177,36 @@ web3.js or a similar module)::
...
-Finally, the functions that will actually end up with the contract and can be called
-by users and contracts alike are ``mint`` and ``send``.
-If ``mint`` is called by anyone except the account that created the contract,
-nothing will happen. This is ensured by the special function ``require`` which
-causes all changes to be reverted if its argument evaluates to false.
-The second call to ``require`` ensures that there will not be too many coins,
-which could cause overflow errors later.
+The functions that make up the contract, and that users and contracts can call are ``mint`` and ``send``.
+
+The ``mint`` function sends an amount of newly created coins from the account that created the contract to another address.
```
Comment (chriseth) [inline docs/introduction-to-smart-contracts.rst]: Also: not from the creator

File: docs/introduction-to-smart-contracts.rst [hunk 8]
```diff
@@ -179,36 +177,36 @@ web3.js or a similar module)::
...
+The :ref:`require <assert-and-require>` function defines conditions that throw an exception if not met, and reverts all changes. In
```
Comment (chriseth) [inline docs/introduction-to-smart-contracts.rst]: I'm not sure we should use the word exception here. Also: The conditions do not throw the exception.

**Comment by chriseth** [comment]
Please squash.

File: docs/introduction-to-smart-contracts.rst (hunk 8 above)
Comment (ChrisChinchilla) [reply docs/introduction-to-smart-contracts.rst]: @chriseth I took the wording from here, so if it needs to change, I am not sure exactly what to, and we should change both.

**Comment by ChrisChinchilla** [comment]
Also, spell checker here is failing on other changes not from this PR

File: docs/introduction-to-smart-contracts.rst (hunk 8 above)
Comment (chriseth) [reply docs/introduction-to-smart-contracts.rst]: We used `exception` in the past but are moving towards `revert` or `failing assertion` now.

File: docs/introduction-to-smart-contracts.rst [hunk 9]
```diff
@@ -115,58 +116,56 @@ registering with username and password — all you need is an Ethereum keypair.
...
-that does not allow any arithmetic operations. It is suitable for
-storing addresses of contracts or of keypairs belonging to external
-persons. The keyword ``public`` automatically generates a function that
-allows you to access the current value of the state variable
-from outside of the contract.
-Without this keyword, other contracts have no way to access the variable.
-The code of the function generated by the compiler is roughly equivalent
+The line ``address public minter;`` declares a state variable of type :ref:`address<address>`.
+The ``address`` type is a 160-bit value that does not allow any arithmetic operations.
+It is suitable for storing addresses of contracts, or a hash of the public half of a keypair belonging to external persons.
```
Comment (chriseth) [inline docs/introduction-to-smart-contracts.rst]: Perhaps we should just go with `addresses of contracts and external accounts`? Maybe the blockchain basics can be linked to explain "external account"?

File: docs/introduction-to-smart-contracts.rst [hunk 10]
```diff
@@ -115,58 +116,56 @@ registering with username and password — all you need is an Ethereum keypair.
...
 Mappings can be seen as `hash tables <https://en.wikipedia.org/wiki/Hash_table>`_ which are
-virtually initialized such that every possible key exists from the start and is mapped to a
-value whose byte-representation is all zeros. This analogy does not go
-too far, though, as it is neither possible to obtain a list of all keys of
-a mapping, nor a list of all values. So either keep in mind (or
-better, keep a list or use a more advanced data type) what you
-added to the mapping or use it in a context where this is not needed.
+virtually initialised such that every possible key exists from the start and is mapped to a
+value whose byte-representation is all zeros. However, it is neither possible to obtain a list of all keys of
+a mapping, nor a list of all values. Keep in mind what you
```
Comment (chriseth) [inline docs/introduction-to-smart-contracts.rst]:
```suggestion
a mapping, nor a list of all values. Either record what you
```
(the developer remembering it does not really help...)

File: docs/introduction-to-smart-contracts.rst [hunk 11]
```diff
@@ -115,58 +116,56 @@ registering with username and password — all you need is an Ethereum keypair.
...
+added to the mapping or use it in a context where this is not needed. Or
+even better, keep a list, or use a more suitable data type.
```
Comment (chriseth) [inline docs/introduction-to-smart-contracts.rst:144]: And then this can be deleted

File: docs/introduction-to-smart-contracts.rst [hunk 12]
```diff
@@ -115,58 +116,56 @@ registering with username and password — all you need is an Ethereum keypair.
...
-a so-called "event" which is emitted in the last line of the function
-``send``. User interfaces (as well as server applications of course) can
-listen for those events being emitted on the blockchain without much
-cost. As soon as it is emitted, the listener will also receive the
-arguments ``from``, ``to`` and ``amount``, which makes it easy to track
-transactions. In order to listen for this event, you would use the following
-JavaScript code (which assumes that ``Coin`` is a contract object created via
-web3.js or a similar module)::
+an :ref:`"event" <events>`, which is emitted in the last line of the function
+``send``. Light Ethereum clients such as web applications can
```
Comment (chriseth) [inline docs/introduction-to-smart-contracts.rst]: Over the years, it turned out that even non-light ethereum clients can not properly function without events, so maybe just change this into "Ethereum applications"?

File: docs/introduction-to-smart-contracts.rst [hunk 13]
```diff
@@ -179,36 +178,33 @@ web3.js or a similar module)::
...
     })
 
-Note how the automatically generated function ``balances`` is called from
-the user interface.
-
 .. index:: coin
 
-The constructor is a special function which is run during creation of the contract and
+The :ref:`constructor<constructor>` is a special function run during the creation of the contract and
 cannot be called afterwards. It permanently stores the address of the person creating the
```
Comment (chriseth) [inline docs/introduction-to-smart-contracts.rst]:
```suggestion
cannot be called afterwards. In this case, it permanently stores the address of the person creating the
```
(to make it clear that this is not automatically the case for all contracts)

File: docs/introduction-to-smart-contracts.rst [hunk 14]
```diff
@@ -179,36 +178,33 @@ web3.js or a similar module)::
...
-nothing will happen. This is ensured by the special function ``require`` which
-causes all changes to be reverted if its argument evaluates to false.
-The second call to ``require`` ensures that there will not be too many coins,
-which could cause overflow errors later.
+The functions that make up the contract, and that users and contracts can call are ``mint`` and ``send``.
+
+The ``mint`` function sends an amount of newly created coins to another address.
+The :ref:`require <assert-and-require>` function defines conditions that throw an exception if not met, and reverts all changes. In
+this example, ``require(msg.sender == minter);`` ensures that only the creator of the contract can call ``mint``,
+and ``require(amount < 1e60);`` ensures a maximum amount of tokens, which could cause overflow errors in the future.
```
Comment (chriseth) [inline docs/introduction-to-smart-contracts.rst]: "The absence of which could cause..."?

Comment (chriseth) [reply docs/introduction-to-smart-contracts.rst]: Or just "without it, we might run into overflow problems"

File: docs/introduction-to-smart-contracts.rst [hunk 15]
```diff
@@ -179,36 +178,33 @@ web3.js or a similar module)::
...
-by users and contracts alike are ``mint`` and ``send``.
-If ``mint`` is called by anyone except the account that created the contract,
-nothing will happen. This is ensured by the special function ``require`` which
-causes all changes to be reverted if its argument evaluates to false.
-The second call to ``require`` ensures that there will not be too many coins,
-which could cause overflow errors later.
+The functions that make up the contract, and that users and contracts can call are ``mint`` and ``send``.
+
+The ``mint`` function sends an amount of newly created coins to another address.
+The :ref:`require <assert-and-require>` function defines conditions that throw an exception if not met, and reverts all changes. In
```
Comment (chriseth) [inline docs/introduction-to-smart-contracts.rst]: Also, the function does not define the condition, the function call does it. But maybe that is pedantic...

**Comment by chriseth** [comment]
Sorry, the more I look at it, the more issues I find, especially in the original version :)

**Comment by ChrisChinchilla** [comment]
@chriseth No worries! That's why I wanted to update it :)

File: docs/introduction-to-smart-contracts.rst (hunk 11 above)
Comment (ChrisChinchilla) [reply docs/introduction-to-smart-contracts.rst:144]: Which part? I feel like at least a part of this explanation is still useful.

File: docs/introduction-to-smart-contracts.rst (hunk 8 above)
Comment (ChrisChinchilla) [reply docs/introduction-to-smart-contracts.rst]: OK, I'll change here and make some future PRs to change this elsewhere.

File: docs/introduction-to-smart-contracts.rst (hunk 11 above)
Comment (chriseth) [reply docs/introduction-to-smart-contracts.rst:144]: The sentence "Or even better, keep a list, ..."

Comment (chriseth) [reply docs/introduction-to-smart-contracts.rst:144]: Actually it is fine as it is.

**Review by chriseth — DISMISSED** [review]

**Comment by chriseth** [comment]
Argh, build is failing: /root/project/docs/introduction-to-smart-contracts.rst:119:undefined label: accounts (if the link has no caption the label must precede a section header)

**Comment by ChrisChinchilla** [comment]
@chriseth Fixed

**Review by chriseth — APPROVED** [review]

**Pull request merged**